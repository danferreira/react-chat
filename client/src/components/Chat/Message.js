import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import styled from 'styled-components';

const MessageWrapper = styled.div`
    word-wrap: break-word;
    font-family: "Open Sans", sans-serif;
    font-size: 20px;
    margin: 10px;

    &:after {
        content: '';
        clear: both;
        display: block;
    }
`;

const MessageBubble = styled.div`
    padding: 10px;
    background: ${props => props.type === 'out'? '#ff0' : '#fff'};
    border-radius: ${props => props.type === 'out'? '10px 10px 0 10px' : '10px 10px 10px 0'};
    float: ${props => props.type === 'out'? 'right' : 'left'};

    @media screen and (max-width: 992px) {
        max-width: 95%;
    }

    @media screen and (min-width: 993px) and (max-width: 1024px) {
        max-width: 85%;
    }
    
    @media screen and (min-width: 1025px) and (max-width: 1300px) {
        max-width: 75%;
    }

    @media screen and (min-width: 1301px) {
        max-width: 65%;
    }
`;

const MessageDate = styled.span`
    color: rgba(0, 0, 0, 0.45);
    font-size: 14px;
`;

const propTypes = {
    message: PropTypes.shape({
        id: PropTypes.number.isRequired,
        content: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired
    }).isRequired
}

const formatDate = date => {

    const mDate = moment(date);
    const mTime = mDate.format('HH:mm');
    const today = moment();
    const yesterday = moment().add(-1, 'days');

    if (mDate.isSame(today, 'day')) {
        return mTime;
    }

    if (mDate.isSame(yesterday, 'day')) {
        return `yesterday at ${mTime}`;
    }

    return `${mDate.format('MM/DD/YYYY')} at ${mTime}`;
}

const Message = ({ message }) =>
    <MessageWrapper>
        <MessageBubble type={message.type}>
            <p>{message.content}</p>
            <MessageDate>{formatDate(message.date)}</MessageDate>
        </MessageBubble>
    </MessageWrapper>

Message.propTypes = propTypes;

export default Message;