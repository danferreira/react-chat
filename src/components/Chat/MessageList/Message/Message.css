.message {
    word-wrap: break-word;
    font-family: "Open Sans", sans-serif;
    font-size: 20px;
    margin: 10px;
}

@media screen and (max-width: 992px) {
    .message-bubble {
        max-width: 95%;
    }
}

@media screen and (min-width: 993px) and (max-width: 1024px) {
    .message-bubble {
        max-width: 85%;
    }
}

@media screen and (min-width: 1025px) and (max-width: 1300px) {
    .message-bubble {
        max-width: 75%;
    }
}

@media screen and (min-width: 1301px) {
    .message-bubble {
        max-width: 65%;
    }
}

.message:after {
    content: '';
    clear: both;
    display: block;
}

.message-out {
    background: #ff0;
    border-radius: 10px 10px 0 10px;
    float: right;
    padding: 10px;
}

.message-in {
    background: #fff;
    border-radius: 10px 10px 10px 0;
    float: left;
    padding: 10px;
}

.message span {
    color: rgba(0, 0, 0, 0.45);
    font-size: 14px;
}