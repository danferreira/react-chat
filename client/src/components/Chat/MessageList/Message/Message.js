import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import moment from 'moment';

import './Message.css';

const propTypes = {
    message: PropTypes.shape({
        id: PropTypes.number.isRequired,
        content: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired
    }).isRequired
}

const formatDate = date => {

    const mDate = moment(date);
    const mTime = mDate.format('HH:mm');
    const today = moment();
    const yesterday = moment().add(-1, 'days');

    if(mDate.isSame(today, 'day')) {
        return mTime;
    }

    if(mDate.isSame(yesterday, 'day')) {
        return `yesterday at ${mTime}`;
    }

    return `${mDate.format('MM/DD/YYYY')} at ${mTime}`;
}

const Message = ({ message }) =>
    <div className="message">
        <div className={classNames("message-bubble", "message-" + message.type)}>
            <p>{message.id + message.content}</p>
            <span>{formatDate(message.date)}</span>
        </div>
    </div>

Message.propTypes = propTypes;

export default Message;