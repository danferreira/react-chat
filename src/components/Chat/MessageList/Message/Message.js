import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import './Message.css';

const propTypes = {
    message: PropTypes.shape({
        id: PropTypes.string.isRequired,
        content: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired
    }).isRequired
}

const Message = ({ message }) =>
    <div className="message">
        <div className={classNames("message-bubble", "message-" + message.type)}>
            <p>{message.content}</p>
            {/* <span>{message.date}</span> */}
        </div>
    </div>

Message.propTypes = propTypes;

export default Message;