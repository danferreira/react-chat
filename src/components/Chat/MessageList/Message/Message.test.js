import React from 'react';
import { shallow } from 'enzyme';

import Message from './Message';

describe('<Message />', () => {

    it('should render an "in" message', () => {
        var message = {
            id: "1",
            content: "hello",
            type: "in"
        };

        var wrapper = shallow(<Message message={message}/>);

        expect(wrapper).toMatchSnapshot();
    })

    it('should render an "out" message', () => {
        var message = {
            id: "1",
            content: "hello",
            type: "out"
        };

        var wrapper = shallow(<Message message={message}/>);

        expect(wrapper).toMatchSnapshot();
    })
});