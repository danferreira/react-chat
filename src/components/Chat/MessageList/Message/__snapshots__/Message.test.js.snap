// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Message /> should render an "in" message 1`] = `
<div
  className="message"
>
  <div
    className="message-bubble message-in"
  >
    <p>
      hello
    </p>
  </div>
</div>
`;

exports[`<Message /> should render an "out" message 1`] = `
<div
  className="message"
>
  <div
    className="message-bubble message-out"
  >
    <p>
      hello
    </p>
  </div>
</div>
`;
